Closes Issue #77. Adds dark theme and modified dependent files.

abhaseen opened the pull request:
<!--
Thanks for sending a pull request!
- if this is your first time, please ensure you have read through our contributor guide: https://github.com/Seneca-CDOT/telescope/blob/master/docs/CONTRIBUTING.md
-->

## Issue This PR Addresses

<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number><reason>
-->

Closes #77 

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description

<!-- Please add a detailed description of what this PR does and why it is needed -->
### Things that were done:
- Moved content from `theme.js` from `telescope/src/frontend/src/` into `TopLayout.js` in `telescope/src/frontend/plugins/gatsby-plugin-top-layout`. This was done in order to be able to use variables to determine the colors based on a light or dark preference.

```js
const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');
```
This determines the color scheme being used by the user preference in the browser/device. If a preference cannot be found, it will default to using a light theme. 
- Moved content and merged into [Material UI hook API style](https://material-ui.com/styles/advanced/) format from `telescope-post-content.css` in `telescope/src/frontend/src/components/Post` into `Post.jsx` in the same folder. This was in order to use the class `telescopePostContent` which will allows us to utilize the variables for dynamic theming.
- In the MobileHeader.jsx file in the Header component, there was a modification made to certain lines of code in order to ensure proper and uniform color scheming of text across mobile and desktop views for the header and the hamburger menu.

### How it looks:
Here are the screenshots of how it looks like:
![TelescopeDarkTheme](https://user-images.githubusercontent.com/34148840/99623819-ef670100-29fa-11eb-920f-c4a2c7738fa7.png)
This is the desktop view for the Dark Mode Scheme
![TelescopeDarkThemeMobile](https://user-images.githubusercontent.com/34148840/99623820-efff9780-29fa-11eb-8d8a-c78f642384ec.png)
This is the mobile view for the Dark Mode Scheme
![TelescopeDarkThemeMobile_w_Navbar](https://user-images.githubusercontent.com/34148840/99623821-efff9780-29fa-11eb-8f07-a07fa84b3d37.png)
This is the mobile view with the Navbar visible (from the hamburger menu)
![TeleScopeLightTheme](https://user-images.githubusercontent.com/34148840/99623822-efff9780-29fa-11eb-8fb8-782bd71d9b82.png)
This is the default light theme for desktop
![TelescopeLightThemeMobile](https://user-images.githubusercontent.com/34148840/99623823-f0982e00-29fa-11eb-8f93-daf4fbbf7772.png)
This is the mobile light theme
![TelescopeLightThemeMobile_w_Navbar](https://user-images.githubusercontent.com/34148840/99623825-f0982e00-29fa-11eb-87fc-365cd2b9a897.png)
This is the mobile light theme with the navbar visible.

## Checklist

<!-- Before submitting a PR, address each item -->

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

agarcia-caicedo reviewed (changes requested):
Hi, @abhaseen, we actually have a dark theme colour palette chosen, which you can find [here](https://github.com/orgs/Seneca-CDOT/teams/telescope-frontend/discussions). We chose those colours as they best fit the material design style. You can read more about that in [this site](https://material.io/design/color/dark-theme.html)

agarcia-caicedo reviewed (changes requested):
The changes to the theme also create a contrast issue here:
![image](https://user-images.githubusercontent.com/49992710/99626068-6d2d0b80-29ff-11eb-96a3-e830e06ca6f9.png)

as #0589D6 and #333E64 do not meet the minimum contrast standard.
![image](https://user-images.githubusercontent.com/49992710/99626198-ad8c8980-29ff-11eb-8484-b503e78be8c0.png)

agarcia-caicedo commented:
Also is there a way an user can chose to toggle this feature?

abhaseen commented:
Hi @agarcia-caicedo! Thank you for the feedback! I just had a few questions/comments:

For the light themes, I was using the current colors in the main branch. The site that you mentioned, uses: #00233a, #e5e5e5, #4f96d8 and #a4d4ff, should I update the light theme colors to these?

For the dark themes, I see you guys chose: #303030, #a4c8e9, #121212 and #4f96d8. I was experimenting with some alternatives but didn't consider the contrast issues. Should I revert to the ones listed here?

There might be some issues regarding certain things like the hyperlinks, active and visited. I currently chose "pink" for dark and "purple" for light. Please let me know if there are better choices.

I was debating as to whether adding the switch but it would impact a bit more code. Should I add a switch? This would likely modify the Header component in the `MobileHeader.jsx` and `DesktopHeader.jsx`.

agarcia-caicedo commented:
@abhaseen don't worry about the light theme, we decided against that theme, and decided to use the ones we currently have in [these docs](https://github.com/Seneca-CDOT/telescope/blob/master/docs/theme-and-colours.md). Sorry this all a bit all over the place. 

You should probably use the suggested colors, as we did test them for contrast. We can always change them, but as a base its a good start.

For hyperlinks, we probably want to reuse the same blue as the scroll button. That means #0589D6 in light mode and #a4c8e9 in dark mode. We haven't had much luck introducing additional hues.

For the switch, I'm ok with opening another issue to implement it. We do need to decide how that would look in terms of design too, since there's different ways to go about it. What do you think @humphd?

abhaseen commented:
Hi @agarcia-caicedo ! As requested, I have made the necessary changes.

# Things That Were Added in the Latest Commit
- Added separated components to store the lightTheme and the darkTheme, this follows a more modular design and this will make it easier to reference for anyone who will work on it in the future. These are all in a new folder called theme.
- Added a theme mapper to select the appropriate theme based on what was toggled. In same folder as mentioned above.
- Modified MobileHeader and DesktopHeader to include a button that will toggle the ability to go from light to dark or vice versa.
- Modified Post to take the text secondary color rather than the secondary.main.
- Modified SearchBar to use textPrimary and default colors rather than the primary.main and primary.contrastText.
- We're now going to be using a button to toggle rather than taking the default user preference. This way users can switch between the two as they please.

# Important changes to color schemes:
The original color in the "Light Theme" that was being used for the post titles based on the scheme you had linked me to was using:
#0589D6 this color was not contrasting well by the standards of the google chrome dev tool suggestions. So I changed it to: #35b4ff which Text Secondary. This was suggested by the Chrome Dev Too:
![LightThemePostTitleOriginal](https://user-images.githubusercontent.com/34148840/99764650-84392f80-2acb-11eb-8874-6da181d41a69.png)
**Above is the original**
![LightThemePostTitle](https://user-images.githubusercontent.com/34148840/99764802-d1b59c80-2acb-11eb-8527-4389d6a6d028.png)
**Above is what it has been changed to - see the right hand side where the Google Chrome Dev Tool suggestion is**

Now, for the links in the light theme, there was also a contrast issue. The original color that was being used was also #0589D6. As suggested by Google Chrome Dev Tool, I changed to #034e7a. See images below:
![LightThemeLinkOriginal](https://user-images.githubusercontent.com/34148840/99765028-3a047e00-2acc-11eb-9c02-a14136577dd7.png)
**Above is the original**
![LightThemeOriginalLink](https://user-images.githubusercontent.com/34148840/99765086-54d6f280-2acc-11eb-821f-8855d568ea68.png)
**Above is the modified version**

And here are the link colors for visited websites:
Pink (#FFC0CB) for the Dark Theme (as seen below):
![DarkThemeLinkVisited](https://user-images.githubusercontent.com/34148840/99765179-7f28b000-2acc-11eb-95ae-38a33fbbd3b6.png)
Purple (#800080) for the Light Theme (as seen below):
![LightThemeLinkVisited](https://user-images.githubusercontent.com/34148840/99765240-a1bac900-2acc-11eb-8bd4-d2748f4eb162.png)

As for the button method, there is a drawback: it does not persist with the session, it will reset to light theme by default upon refresh.

humphd commented:
This needs a rebase for conflicts in src/frontend/src/components/SearchBar/SearchBar.jsx

manekenpix commented:
As @humphd, this needs a rebase, and please don't merge master into your branch:
![image](https://user-images.githubusercontent.com/23108901/99864964-74c1f100-2b74-11eb-8168-99e4d4d89e1c.png)

agarcia-caicedo commented:
Accidentally approved, rebase needs to happen, but I do like the changes

agarcia-caicedo reviewed (changes requested):
The search page has some issues: 
Both the placeholder text and the text in the search field are dark in dark mode:

![image](https://user-images.githubusercontent.com/49992710/99865220-388f9000-2b76-11eb-9afa-7c29c781ab85.png)

abhaseen commented:
Hi! @agarcia-caicedo . Worked on a few things:
- I addressed the About and Search not showing up properly.
- I think I rebased the repo properly... Correct me if I'm wrong, I might need some guidance with this.

Here are some screenshots:
![Screenshot from 2020-11-21 20-57-54](https://user-images.githubusercontent.com/34148840/99891980-e6b23d00-2c3d-11eb-9736-ef1357058e11.png)
**Above is the Dark Theme Search Page** I fixed the invisible placeholder and the results text color, since it wasn't using the default color before.
![Screenshot from 2020-11-21 20-58-05](https://user-images.githubusercontent.com/34148840/99891993-13feeb00-2c3e-11eb-8f1c-c3c283e66d85.png)
**Above is the Light Theme Search Page** I fixed the results text color, since it wasn't using the default color before.
![Screenshot from 2020-11-21 20-58-10](https://user-images.githubusercontent.com/34148840/99892009-2d079c00-2c3e-11eb-80dd-7aa0d83ce4a4.png)
**Above is the light Theme About Page** Fixed it to display proper colors for text AND links.
![Screenshot from 2020-11-21 20-58-14](https://user-images.githubusercontent.com/34148840/99892021-4ad50100-2c3e-11eb-9934-a0568509d5c6.png)
**Above is the Dark Theme About Page** Fixed it to display the proper color for the links.

agarcia-caicedo commented:
Looks good so far to me. There are two pages I currently can't check - 404 and myfeeds. Myfeeds hasn't been fully converted to light mode, so it might have issues or it might not. 

You might want to check those first, but other than that I'm ready to approve the changes

abhaseen commented:
@agarcia-caicedo , I updated the error/404 pages in "error.jsx" to match the styling palette we are using. Specifically, I changed the background color & color tags of the button.

Here are the screenshots:
![Screenshot from 2020-11-22 01-16-27](https://user-images.githubusercontent.com/34148840/99896746-83d29d00-2c61-11eb-8ad9-c5a5c7152127.png)
**Above is the light mode page**
![Screenshot from 2020-11-22 01-16-25](https://user-images.githubusercontent.com/34148840/99896753-8fbe5f00-2c61-11eb-9770-619547e09455.png)
**Above is the dark mode page**

I have no way of verifying the myfeeds, so someone will have to take a look at that and make any necessary changes as needed.

humphd commented:
@tonyvugithub can you look at this from a theme-perspective for me?

tonyvugithub commented on src/frontend/plugins/gatsby-plugin-top-layout/TopLayout.js:
```diff
@@ -1,10 +1,44 @@
+    },
+  });
+
   return (
```
@humphd : This is great work but I am afraid that it will create more work when we port to Next. To minimize the amount of work in Next, the easiest way is to follow MUI theming structure which was what i did in #1397 .  
![image](https://user-images.githubusercontent.com/55034244/99472026-c6207500-2915-11eb-8cf8-653eaca28424.png)

tonyvugithub replied:
the `text` property again will be a problem when we port to NextJS

tonyvugithub replied:
There is a workaround that allows you to kinda overwrite default MUI property. In that case, you can do whatever structure you like. The main culprit here is TypeScript. It is quite strict.

tonyvugithub replied:
@abhaseen

tonyvugithub commented:
@humphd : Do we have a final agreement to what approach we should do? because this will affect my #1386 theming set up for Next and I am going to start on it shortly.

humphd commented:
I don't know enough about theming to answer this, except to say that I'd like to do what's going to work for our stack: Next.js, TypeScript, and Material-UI.  Whatever those three want, that's what we need to do.

tonyvugithub commented:
@abhaseen : I might have to ask you to switch your theme.js file to its previous state and if make any changes, please refer to the MUI theming structure that I have included in one of the comments. The sooner you can resolve this, the sooner we can port theming in NextJS

abhaseen commented:
Hi @tonyvugithub ! Sorry I haven't been replying, I just wanted to let everyone know I'm still working on this, but I won't be able to push any meaningful updates for the next couple of weeks (due to exams, I kind of want to give this my undivided attention). Just a brief question: for the customs themes that I implemented, would you want me to follow the scheme similar to what you have posted? I had seen you do some work in PR #1386 , so I just wanted to know if I should work with those files that have been set up. Please let me know, thanks!

tonyvugithub commented:
@abhaseen : hi, it would be easier for porting if your theme can follow the same structure that I have in my issue. Later on we can come back to this and adjust. During the posting process, it is better to stick to the default option as much as possible. I really appreciate your cooperation. Thanks @abhaseen , good luck with your exams.

yuanLeeMidori commented:
@abhaseen Hi! Thank you for the great work :+1: Telescope team is handling the open PRs for the upcoming release. In that case, we need to know if you're still working on this PR. We're happy to see you continue the work; but if you don't want to keep working on it, we'll be assigning this PR to other team members.
Thank you! The work you have done is impressive.

abhaseen commented:
@yuanLeeMidori , yes, I am still working on it, sorry I've been a bit behind on releases because I've been occupied with other projects, but I do plan on doing one more push soon.

yuanLeeMidori commented:
@abhaseen no worries, we just need to know if you're still working on it. Please reach out on Slack if you need any help :)

tonyvugithub commented:
@abhaseen : Thanks for your interest to continue. We decided to push this to 1.6 release and skip it for 1.5 release since the files have changed and we want to finished the migration to Next first before making any changes to the styling or components. We keep you posted on the progress. Thanks again

abhaseen commented:
Hi. So after working on this for a while and while I was doing the rebase, I realized that there are too many frustrating manual edits to make since the last updates. My suggestion going forward for anyone who wants to take this up would be to create a new PR. The themeing and colors should be used for reference. But the code is a bit too all over the place in comparison to the current state of the code as of Update v.1.5. I think it would be best if I hand this off to someone who is more familiar with the Next.JS coding that's been done since the last several updates. Sorry I could not see this through. I wish you all the best of luck in moving this forward.

tonyvugithub commented:
Closed in favour of #1637
